File: src/domain/budget.ts
import { addMoney, rateOfMoney, subtractMoney, type Halalas } from './money'
import type { Id } from './entities/types'

/**
 * الميزانية والمتوسطات والشذوذ — spec/01 و spec/06.
 *
 * القاعدتان الحاكمتان في هذا الملف:
 *
 * ١. «**لا تُخلق ميزانيات من متوسطات دون اختيار المستخدم**» (spec/01).
 *    المتوسط معلومة تُعرض، لا سقف يُطبَّق. الاثنان لا يختلطان أبدًا.
 *
 * ٢. «تصنيف بلا سقف/تاريخ ⇒ **لا سقف أو متوسط مخترع ولا تنبيه عتبة**» (spec/06).
 *    فكل دالة هنا ترجّع `null` وسببًا حين لا يكفي الدليل.
 */

/* ───────────────────────── حالة السقف ───────────────────────── */

export type BudgetLevel = 'under' | 'near' | 'over'

export interface BudgetStatus {
  limitMinor: Halalas
  spentMinor: Halalas
  /** المتبقي؛ سالب حين يُتجاوز السقف — ولا يُخفى (spec/02). */
  remainingMinor: Halalas
  /** النسبة المستهلكة بالعُشر من المئة (855 = 85.5٪). عدد صحيح. */
  usedTenthPercent: number
  level: BudgetLevel
  /** هل تجاوزت عتبة التنبيه التي حددها المستخدم؟ */
  thresholdCrossed: boolean
}

/**
 * حالة سقف واحد.
 *
 * `thresholdPercent` عتبة **يحددها المستخدم**؛ بدونها لا تنبيه (spec/06).
 */
export function budgetStatus(
  limitMinor: Halalas,
  spentMinor: Halalas,
  thresholdPercent: number | null,
): BudgetStatus {
  if (limitMinor <= 0) throw new Error('السقف لازم يكون أكبر من صفر')

  const usedTenthPercent = rateOfMoney(spentMinor, 1000, limitMinor)
  const remainingMinor = subtractMoney(limitMinor, spentMinor)

  const level: BudgetLevel =
    usedTenthPercent >= 1000 ? 'over' : usedTenthPercent >= 900 ? 'near' : 'under'

  return {
    limitMinor,
    spentMinor,
    remainingMinor,
    usedTenthPercent,
    level,
    thresholdCrossed:
      thresholdPercent === null ? false : usedTenthPercent >= thresholdPercent * 10,
  }
}

/* ───────────────────────── المتوسط ───────────────────────── */

/** أقل عدد فترات مكتملة قبل أن يكون للمتوسط معنى. */
export const MIN_PERIODS_FOR_AVERAGE = 3

/** فترة مضت، بمصروفها ومدى موثوقيته. */
export interface CompletedPeriodSpend {
  periodKey: string
  spentMinor: Halalas
  /**
   * هل كل عمليات الفترة محددة النوع الاقتصادي؟
   * فترة نصف محددة **لا تدخل المتوسط**: مصروفها معروف جزئيًا،
   * وإدخاله يجعل المتوسط أقل من الحقيقة بلا أن يظهر ذلك.
   */
  reliable: boolean
  transactionCount: number
}

export interface AverageResult {
  /** null = البيانات لا تكفي. لا يُخترع متوسط. */
  averageMinor: Halalas | null
  /** الفترات التي دخلت الحساب فعلًا. */
  usedPeriods: string[]
  /** الفترات المستبعدة وسبب استبعادها — يُعرض للمستخدم لا يُخفى. */
  excluded: { periodKey: string; reason: string }[]
  reason: string
}

/**
 * متوسط المصروف من **الفترات المكتملة الموثوقة فقط**.
 *
 * spec/01: «متوسط منفصل من فترات مكتملة».
 * الفترة الجارية مستبعدة دائمًا: مصروفها ناقص بطبيعته، وإدخاله
 * يسحب المتوسط لأسفل كل يوم.
 */
export function averageCompletedSpend(
  periods: readonly CompletedPeriodSpend[],
): AverageResult {
  const excluded: { periodKey: string; reason: string }[] = []
  const used: CompletedPeriodSpend[] = []

  for (const p of periods) {
    if (!p.reliable) {
      excluded.push({
        periodKey: p.periodKey,
        reason: 'فيها عمليات لسه محتاجة تحديد نوعها، فمصروفها ناقص',
      })
      continue
    }
    if (p.transactionCount === 0) {
      excluded.push({
        periodKey: p.periodKey,
        reason: 'مفيش فيها عمليات — مش واضح ده صفر حقيقي ولا بيانات ناقصة',
      })
      continue
    }
    used.push(p)
  }

  if (used.length < MIN_PERIODS_FOR_AVERAGE) {
    return {
      averageMinor: null,
      usedPeriods: used.map((p) => p.periodKey),
      excluded,
      reason:
        `المتوسط محتاج ${MIN_PERIODS_FOR_AVERAGE} فترات مكتملة على الأقل، ` +
        `والمتاح ${used.length}.`,
    }
  }

  const total = addMoney(...used.map((p) => p.spentMinor))
  return {
    averageMinor: Math.round(total / used.length),
    usedPeriods: used.map((p) => p.periodKey),
    excluded,
    reason: `متوسط ${used.length} فترة مكتملة.`,
  }
}

/* ───────────────────────── الشذوذ ───────────────────────── */

/** أقل عدد قيم تاريخية قبل ادعاء شذوذ. */
export const MIN_HISTORY_FOR_ANOMALY = 3

/**
 * الحد الأدنى للانحراف النسبي قبل الوصف بالشذوذ: ٢٠٪.
 * يمنع وسم فرق تافه بأنه شذوذ لمجرد أن التاريخ متطابق.
 */
const MIN_RELATIVE_DEVIATION_PER_THOUSAND = 200

/** مضاعف الانحراف المطلق المتوسط الذي يُعد بعده الرقم شاذًا. */
const ROBUST_MULTIPLIER = 3

export interface AnomalyResult {
  /** null = لا يمكن الحكم. **ليست «لا يوجد شذوذ»**. */
  isAnomaly: boolean | null
  /** الوسيط التاريخي، للعرض والمقارنة. */
  medianMinor: Halalas | null
  /** الفرق عن الوسيط؛ موجب = أعلى. */
  deviationMinor: Halalas | null
  reason: string
}

function median(values: readonly number[]): number {
  const sorted = [...values].sort((a, b) => a - b)
  const mid = Math.floor(sorted.length / 2)
  return sorted.length % 2 === 1
    ? sorted[mid]
    : Math.round((sorted[mid - 1] + sorted[mid]) / 2)
}

/**
 * يكشف الشذوذ بمقارنة قيمة بالوسيط والانحراف المطلق الوسيط (MAD).
 *
 * لماذا الوسيط لا المتوسط: شهر واحد استثنائي (سفر مثلًا) يرفع المتوسط
 * فيخفي شذوذ الشهر التالي. الوسيط لا يتأثر بقيمة شاذة واحدة.
 *
 * **لا يُدّعى شذوذ بلا تاريخ كافٍ**، والنتيجة `null` لا `false`:
 * «ما نعرفش» غير «مفيش شذوذ» (spec/06: لا سقف ولا متوسط مخترع).
 */
export function detectAnomaly(
  valueMinor: Halalas,
  historyMinor: readonly Halalas[],
): AnomalyResult {
  if (historyMinor.length < MIN_HISTORY_FOR_ANOMALY) {
    return {
      isAnomaly: null,
      medianMinor: null,
      deviationMinor: null,
      reason:
        `الحكم على الشذوذ محتاج ${MIN_HISTORY_FOR_ANOMALY} فترات سابقة على الأقل، ` +
        `والمتاح ${historyMinor.length}.`,
    }
  }

  const med = median(historyMinor)
  const deviation = valueMinor - med
  const absDeviation = Math.abs(deviation)

  if (med === 0) {
    return {
      isAnomaly: null,
      medianMinor: 0,
      deviationMinor: deviation,
      reason: 'الوسيط التاريخي صفر، فالمقارنة النسبية مالهاش معنى.',
    }
  }

  // الانحراف لازم يكون كبيرًا نسبيًا **وأيضًا** شاذًا مقارنة بتقلب التاريخ
  const relativePerThousand = rateOfMoney(absDeviation, 1000, Math.abs(med))
  if (relativePerThousand < MIN_RELATIVE_DEVIATION_PER_THOUSAND) {
    return {
      isAnomaly: false,
      medianMinor: med,
      deviationMinor: deviation,
      reason: `الفرق عن المعتاد أقل من ٢٠٪، فمش شذوذ.`,
    }
  }

  const mad = median(historyMinor.map((v) => Math.abs(v - med)))

  // تاريخ ثابت تمامًا: التقلب صفر، فأي انحراف نسبي كبير يُعد شذوذًا
  const isAnomaly = mad === 0 ? true : absDeviation >= mad * ROBUST_MULTIPLIER

  const direction = deviation > 0 ? 'أعلى' : 'أقل'
  const percent = (relativePerThousand / 10).toFixed(0)

  return {
    isAnomaly,
    medianMinor: med,
    deviationMinor: deviation,
    reason: isAnomaly
      ? `${direction} من المعتاد بـ${percent}٪ — خارج تقلب الفترات السابقة.`
      : `${direction} من المعتاد بـ${percent}٪، لكن ده داخل تقلب الفترات السابقة العادي.`,
  }
}

/* ───────────────────────── ربط السقوف بالتصنيفات ───────────────────────── */

export interface CategoryBudgetLine {
  categoryId: Id
  /** null = المستخدم لم يضع سقفًا لهذا التصنيف. */
  status: BudgetStatus | null
  spentMinor: Halalas
  averageMinor: Halalas | null
  anomaly: AnomalyResult
  /** سبب غياب السقف — يُعرض ولا يُترك فراغًا. */
  noLimitReason: string | null
}

/**
 * يبني سطور شاشة الميزانية.
 *
 * التصنيف بلا سقف **يُعرض بمصروفه ومتوسطه**، لكن بلا حالة سقف
 * وبلا أي تنبيه عتبة — spec/06.
 */
export function buildCategoryLines(
  spendByCategory: ReadonlyMap<Id, Halalas>,
  limitByCategory: ReadonlyMap<Id, { limitMinor: Halalas; thresholdPercent: number | null }>,
  averageByCategory: ReadonlyMap<Id, Halalas | null>,
  historyByCategory: ReadonlyMap<Id, readonly Halalas[]>,
): CategoryBudgetLine[] {
  const ids = new Set<Id>([...spendByCategory.keys(), ...limitByCategory.keys()])

  return [...ids]
    .map((categoryId) => {
      const spentMinor = spendByCategory.get(categoryId) ?? 0
      const limit = limitByCategory.get(categoryId)
      return {
        categoryId,
        spentMinor,
        status: limit ? budgetStatus(limit.limitMinor, spentMinor, limit.thresholdPercent) : null,
        averageMinor: averageByCategory.get(categoryId) ?? null,
        anomaly: detectAnomaly(spentMinor, historyByCategory.get(categoryId) ?? []),
        noLimitReason: limit ? null : 'مفيش سقف للتصنيف ده. تقدر تحدده من هنا.',
      }
    })
    .sort((a, b) => b.spentMinor - a.spentMinor)
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/Budget.kt
package app.masroufy.core

import kotlin.math.abs

/**
 * الميزانية والمتوسطات والشذوذ — نقل `src/domain/budget.ts` (spec/01 و spec/06).
 * ١. **مفيش ميزانية بتتعمل من متوسط من غير اختيار المستخدم** — المتوسط معلومة مش سقف.
 * ٢. تصنيف من غير سقف أو تاريخ ⇒ **مفيش سقف ولا متوسط مخترع ولا تنبيه** — null وسبب.
 */
enum class BudgetLevel(val wire: String) { UNDER("under"), NEAR("near"), OVER("over") }

data class BudgetStatus(
    val limitMinor: Halalas,
    val spentMinor: Halalas,
    /** سالب لما السقف يتعدّى — وما بيتخباش. */
    val remainingMinor: Halalas,
    /** المستهلك بالعُشر من المية (855 = 85.5٪). */
    val usedTenthPercent: Long,
    val level: BudgetLevel,
    /** عدّى عتبة التنبيه اللي حددها المستخدم؟ */
    val thresholdCrossed: Boolean,
)

class BudgetError(message: String) : IllegalArgumentException(message)

fun budgetStatus(limitMinor: Halalas, spentMinor: Halalas, thresholdPercent: Int?): BudgetStatus {
    if (limitMinor <= 0) throw BudgetError(uiText(TextKey.BUDGET_LIMIT_POSITIVE))
    val used = rateOfMoney(spentMinor, 1000, limitMinor)
    val level = when {
        used >= 1000 -> BudgetLevel.OVER
        used >= 900 -> BudgetLevel.NEAR
        else -> BudgetLevel.UNDER
    }
    return BudgetStatus(limitMinor, spentMinor, subtractMoney(limitMinor, spentMinor), used, level, thresholdPercent != null && used >= thresholdPercent * 10L)
}

const val MIN_PERIODS_FOR_AVERAGE = 3

data class CompletedPeriodSpend(
    val periodKey: String,
    val spentMinor: Halalas,
    /** كل عمليات الفترة نوعها محدد؟ الفترة الناقصة ما بتدخلش المتوسط. */
    val reliable: Boolean,
    val transactionCount: Int,
)

data class ExcludedPeriod(val periodKey: String, val reason: String)

data class AverageResult(
    /** null = البيانات ما تكفيش. */
    val averageMinor: Halalas?,
    val usedPeriods: List<String>,
    val excluded: List<ExcludedPeriod>,
    val reason: String,
)

/** متوسط **الفترات المكتملة الموثوقة بس** — الفترة الجارية برا دايمًا. */
fun averageCompletedSpend(periods: List<CompletedPeriodSpend>): AverageResult {
    val excluded = mutableListOf<ExcludedPeriod>()
    val used = mutableListOf<CompletedPeriodSpend>()
    for (p in periods) {
        when {
            !p.reliable -> excluded += ExcludedPeriod(p.periodKey, uiText(TextKey.PERIOD_EXCLUDED_UNRELIABLE))
            p.transactionCount == 0 -> excluded += ExcludedPeriod(p.periodKey, uiText(TextKey.PERIOD_EXCLUDED_EMPTY))
            else -> used += p
        }
    }
    if (used.size < MIN_PERIODS_FOR_AVERAGE) {
        return AverageResult(
            null, used.map { it.periodKey }, excluded,
            uiText(TextKey.AVERAGE_NEEDS_PERIODS, MIN_PERIODS_FOR_AVERAGE.toString(), used.size.toString()),
        )
    }
    val total = sumMoney(used.map { it.spentMinor })
    return AverageResult(JsText.round(total.toDouble() / used.size).toLong(), used.map { it.periodKey }, excluded, uiText(TextKey.AVERAGE_OF_PERIODS, used.size.toString()))
}

const val MIN_HISTORY_FOR_ANOMALY = 3
private const val MIN_RELATIVE_DEVIATION_PER_THOUSAND = 200L
private const val ROBUST_MULTIPLIER = 3L

data class AnomalyResult(
    /** null = ما نقدرش نحكم — **مش** «مفيش شذوذ». */
    val isAnomaly: Boolean?,
    val medianMinor: Halalas?,
    /** موجب = أعلى من المعتاد. */
    val deviationMinor: Halalas?,
    val reason: String,
)

private fun median(values: List<Long>): Long {
    val sorted = values.sorted()
    val mid = sorted.size / 2
    return if (sorted.size % 2 == 1) sorted[mid] else JsText.round((sorted[mid - 1] + sorted[mid]).toDouble() / 2).toLong()
}

/** الشذوذ بالوسيط والانحراف المطلق الوسيط — شهر استثنائي واحد ما يخبيش شذوذ اللي بعده. */
fun detectAnomaly(valueMinor: Halalas, historyMinor: List<Halalas>): AnomalyResult {
    if (historyMinor.size < MIN_HISTORY_FOR_ANOMALY) {
        return AnomalyResult(
            null, null, null,
            uiText(TextKey.ANOMALY_NEEDS_HISTORY, MIN_HISTORY_FOR_ANOMALY.toString(), historyMinor.size.toString()),
        )
    }
    val med = median(historyMinor)
    val deviation = valueMinor - med
    val absDeviation = abs(deviation)
    if (med == 0L) return AnomalyResult(null, 0, deviation, uiText(TextKey.ANOMALY_MEDIAN_ZERO))
    val relative = rateOfMoney(absDeviation, 1000, abs(med))
    if (relative < MIN_RELATIVE_DEVIATION_PER_THOUSAND) {
        return AnomalyResult(false, med, deviation, uiText(TextKey.ANOMALY_WITHIN_TWENTY))
    }
    val mad = median(historyMinor.map { abs(it - med) })
    val isAnomaly = mad == 0L || absDeviation >= mad * ROBUST_MULTIPLIER
    val direction = if (deviation > 0) uiText(TextKey.ANOMALY_HIGHER) else uiText(TextKey.ANOMALY_LOWER)
    // نفس `(n / 10).toFixed(0)`: قيمة بخانة عشرية واحدة، والنص بيروح لفوق
    val percent = (relative + 5) / 10
    return AnomalyResult(
        isAnomaly, med, deviation,
        if (isAnomaly) uiText(TextKey.ANOMALY_OUTSIDE, direction, percent.toString())
        else uiText(TextKey.ANOMALY_INSIDE, direction, percent.toString()),
    )
}

data class CategoryLimit(val limitMinor: Halalas, val thresholdPercent: Int?)

data class CategoryBudgetLine(
    val categoryId: Id,
    /** null = المستخدم ما حطش سقف. */
    val status: BudgetStatus?,
    val spentMinor: Halalas,
    /** نصيبه من مجموع مصروف التصنيفات بالعُشر في المية (OVERRIDES §31). */
    val shareTenthPercent: Long,
    val averageMinor: Halalas?,
    val anomaly: AnomalyResult,
    /** سبب غياب السقف — بيتعرض مش بيتساب فاضي. */
    val noLimitReason: String?,
)

/** سطور شاشة الميزانية — اللي من غير سقف بيتعرض بمصروفه ومتوسطه، من غير حالة سقف ولا تنبيه. */
fun buildCategoryLines(
    spendByCategory: Map<Id, Halalas>,
    limitByCategory: Map<Id, CategoryLimit>,
    averageByCategory: Map<Id, Halalas?>,
    historyByCategory: Map<Id, List<Halalas>>,
): List<CategoryBudgetLine> {
    val ids = LinkedHashSet(spendByCategory.keys).apply { addAll(limitByCategory.keys) }
    var total = 0L
    for (spent in spendByCategory.values) if (spent > 0) total = addMoney(total, spent)
    return ids.map { id ->
        val spent = spendByCategory[id] ?: 0L
        val limit = limitByCategory[id]
        CategoryBudgetLine(
            categoryId = id,
            status = limit?.let { budgetStatus(it.limitMinor, spent, it.thresholdPercent) },
            spentMinor = spent,
            shareTenthPercent = if (total > 0 && spent > 0) rateOfMoney(spent, 1000, total) else 0,
            averageMinor = averageByCategory[id],
            anomaly = detectAnomaly(spent, historyByCategory[id].orEmpty()),
            noLimitReason = if (limit != null) null else uiText(TextKey.CATEGORY_NO_LIMIT),
        )
    }.sortedByDescending { it.spentMinor }
}

/** ارتفاع أعمدة «آخر ست فترات» بالعُشر في المية من أكبر فترة (OVERRIDES §31) — null بيفضل null. */
fun relativeTenths(values: List<Halalas?>): List<Long?> {
    val max = values.fold(0L) { highest, v -> if (v != null && v > highest) v else highest }
    return values.map { v ->
        when {
            v == null -> null
            max <= 0 || v <= 0 -> 0
            else -> rateOfMoney(v, 1000, max)
        }
    }
}
